### PR TITLE
MGDOBR-985: Display the Error Handling configuration of a bridge

### PR DESCRIPTION
**Pull request by vpellegrino**
**Task**: https://issues.redhat.com/browse/MGDOBR-985

image

In this Task, you will also find:
- Externalizing `InstancePage`'s tabs content to two different page components
- Load tab content only when accessing to it
- Each tab has a specific route

### CODE REVIEW

File: src/app/Instance/InstancePage/InstancePage.tsx [hunk 1]
```diff
@@ -158,6 +177,29 @@ const InstancePage = (): JSX.Element => {
     }
   }, [bridge, bridgeError, getPageTitle, history, processorsError, t]);
 
+  useEffect(() => {
+    if (bridge?.error_handler?.type) {
```
Comment (manstis) [inline src/app/Instance/InstancePage/InstancePage.tsx:181]: This is likely to need refactoring when @kostola merges https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1069

His PR adds a new [`type`](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1069/files#diff-66557df441afc3933a43c346564ae30380002fb47b8a0f12612ebcf2e71ece79R307) that has no User configurable parameters and is effectively a marker to say "we handle it all internally".

I added `frontend` reviewers to his PR.. but you might like to create a new JIRA too.

**Review by manstis — APPROVED** [review]

**Comment by vpellegrino** [comment]
As stated in the description of the task, the tab split + subrouting should be managed in this task.
So temporarily closing it.

**Review by jomarko — COMMENTED** [review]
@vpellegrino thank you for a PR.

Is the [pictrue](https://user-images.githubusercontent.com/22591802/186874975-453fe64e-85cb-43ab-92ea-2294abe32d62.png) from PR description up to date? I mean is that the latest version fo your changes?
I have these questions:
- why there is sample password at a place for username?
- why there is sample username at a place of apssword?
- is the password in readable form by design? during creating processors I think we have passwords masked with asterisks 


If the picture is not latest version, feel free to ignore my questions and I will re-review.

**Comment by vpellegrino** [comment]
@jomarko I inverted the user and password for the mocked data. You can review it now.

File: src/app/Instance/InstancePage/InstancePage.tsx (hunk 1 above)
Comment (vpellegrino) [reply src/app/Instance/InstancePage/InstancePage.tsx]: Created a new task also for the UI: https://issues.redhat.com/browse/MGDOBR-1056

**Comment by kelvah** [comment]
Hi @vpellegrino, thanks for taking care of this!

A couple of observations before reviewing.

When I click on the "Error handling" tab and then I click on the "Processors" one again, the links inside the page stop working (create processor, single processors etc.). As a side note, I see you opted for adding top level routes instead of [nesting them](https://v5.reactrouter.com/web/example/nesting). Not sure if the second one may have helped with managing relative links?

Usually when splitting pages I go with getting rid of tabs and adding an horizontal navigation instead ([PF example here](https://www.patternfly.org/v4/components/navigation/design-guidelines#secondary-horizontal-navigation)). Not saying we can't still use tabs, only asking if you considered the option? Could it be simpler?

Last, It seems that the layout of the page content changed a bit, there's no more gray background and white padded content. Maybe some `PageSection`s got removed during the refactoring.

**Comment by vpellegrino** [comment]
Hi @kelvah thank you very much for noticing me about these two issues.

The instance detail page (processors tab) links are now fixed.
The usage of `location.pathname` was not useful anymore in our case because with the introduction of `/instance/<instance_id>/processors` and `/instance/<instance_id>/error-handling`, the `location.pathname` was keeping these intermediary paths and appending the `create-processor` and the link to the processor itself. Honestly, I don't see how nested routing would help in this case, but I am open if you have a working solution (feel free to open a PR against my branch).

Regarding horizontal navigation vs tabs, I don't have a strong opinion.

About the page layout, definitely yes, I was missing a `PageSection` during refactoring. Good catch!

**Comment by kelvah** [comment]
> Honestly, I don't see how nested routing would help in this case, but I am open if you have a working solution (feel free to open a PR against my branch).

I've quickly [put together a PR](https://github.com/vpellegrino/sandbox-ui/pull/5/files) with nested routes and sub-nav to try them out.
Not sure if they are an improvement or not. Maybe they are just an alternative way of doing the same thing.
Waiting for your opinion :slightly_smiling_face:

**Comment by jomarko** [comment]
### 01 tooltips
- During creation of processor, we have tooltips as questionmark icons next to label
- During error handler details viewing, we have tooltips as links dirrctly on label text

Is that intentional?

**Comment by jomarko** [comment]
### 02 password
Even if I didn't configured password for webhook error handler, in details screen it seems like a password was configured. Is that expected?
Screenshot from 2022-09-02 10-06-13

**Comment by vpellegrino** [comment]
> ### 01 tooltips
> * During creation of processor, we have tooltips as questionmark icons next to label
> * During error handler details viewing, we have tooltips as links dirrctly on label text
> 
> Is that intentional?

This is the same behavior as in the processor creation / detail page.

**Comment by vpellegrino** [comment]
> ### 02 password
> Even if I didn't configured password for webhook error handler, in details screen it seems like a password was configured. Is that expected? Screenshot from 2022-09-02 10-06-13

It is not possible to create a webhook error handler w/o a password:
image

Probably, a browser-saved password has been used. Please, check again.
There is a PR about this topic: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/159

**Comment by vpellegrino** [comment]
> > Honestly, I don't see how nested routing would help in this case, but I am open if you have a working solution (feel free to open a PR against my branch).
> 
> I've quickly [put together a PR](https://github.com/vpellegrino/sandbox-ui/pull/5) with nested routes and sub-nav to try them out. Not sure if they are an improvement or not. Maybe they are just an alternative way of doing the same thing. Waiting for your opinion slightly_smiling_face

Hi @kelvah, thank you very much. 
Comparing our solutions is always a source of growth for me.
I answered there: https://github.com/vpellegrino/sandbox-ui/pull/5#issuecomment-1235236653

**Comment by jomarko** [comment]
https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235223755 I do not agree

### processor creating
Screenshot from 2022-09-02 11-14-49


### eh details
Screenshot from 2022-09-02 11-14-58

**Comment by vpellegrino** [comment]
> [#145 (comment)](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235223755) I do not agree
> 
> ### processor creating
> Screenshot from 2022-09-02 11-14-49
> 
> ### eh details
> Screenshot from 2022-09-02 11-14-58

@jomarko, you are comparing editing (processor creating) vs visualization (eh detail).
As I already stated, the same behavior is in place for the processor creation/edit vs processor visualization.

**Comment by jomarko** [comment]
https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235225791

I can not confirm this, even in incognito widnow, it allows me to create webhook eh just by providing endpoint, I aws not prompted to provide username or pasword

**Comment by jomarko** [comment]
https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235272208

I probably do not understand, why we do not use consistent approach. Is that from UX team design?

**Comment by vpellegrino** [comment]
> [#145 (comment)](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235225791)
> 
> I can not confirm this, even in incognito widnow, it allows me to create webhook eh just by providing endpoint, I aws not prompted to provide username or password

So, validation is not working? Can you please send a recording (video or animated GIF)?

**Comment by vpellegrino** [comment]
> [#145 (comment)](https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145#issuecomment-1235272208)
> 
> I probably do not understand, why we do not use consistent approach. Is that from UX team design?

This behavior was already in place before this PR.
I think that it is something coming from Uniforms, but I am not totally sure (maybe @kelvah can confirm).
If you think that this inconsistent approach can lead to issues for customers, I suggest you raise this point with the UX team during the next meeting.

**Comment by jomarko** [comment]
@vpellegrino form validation, that allowed me proceed even without username https://issues.redhat.com/browse/MGDOBR-985?focusedCommentId=20889688&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20889688

**Comment by vpellegrino** [comment]
> @vpellegrino form validation, that allowed me proceed even without username https://issues.redhat.com/browse/MGDOBR-985?focusedCommentId=20889688&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-20889688

@jomarko, thanks for reporting it.
The issue is about the EH validation during the Smart Events instance creation. Can you open a JIRA for it?

About this PR, EH configuration visualization, I think the page is behaving well because, with the asterisks, it is disclosing information about the password. However, that behavior is related to how Uniforms render fields.

**Review by jomarko — APPROVED** [review]
Thank you @vpellegrino I will add cypress tests once this is merged https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/169

**Comment by kelvah** [comment]
Hi @jomarko, I'll try to answer some of your questions.

> 01 tooltips
> During creation of processor, we have tooltips as questionmark icons next to label
> During error handler details viewing, we have tooltips as links dirrctly on label text

This is just how PF handles helper texts. Within forms they are displayed as tooltips on the label icon ([example](https://www.patternfly.org/v4/components/form#basic)), while when showing data inside a page, they are shown by clicking on the data property name ([example](https://www.patternfly.org/v4/components/description-list/#term-help-text)). We are just using PF components and following their designs.

> 02 password
> Even if I didn't configured password for webhook error handler, in details screen it seems like a password was configured. Is that expected?

That shouldn't happen. I think you've found a bug! This is also happening while creating processors and viewing the processor detail page.
Optional secret fields that have no value should not be displayed with asterisks but with "Property not configured". [I've opened a ticket](https://issues.redhat.com/browse/MGDOBR-1059) cause this issue was not introduced by this PR.

**Review by kelvah — COMMENTED** [review]
Hi @vpellegrino, I've left some observations for you. Thanks.

File: src/app/Instance/InstancePage/InstancePage.tsx [hunk 2]
```diff
@@ -256,210 +136,146 @@ const InstancePage = (): JSX.Element => {
...
+                    ]}
                   />
-                </Tabs>
+                </SplitItem>
+              </Split>
+            </PageSection>
+          </>
+        )}
+        <PageSection variant={PageSectionVariants.light} type="tabs">
+          <Tabs
```
[the remark this answers is not in the document]
Comment (vpellegrino) [reply src/app/Instance/InstancePage/InstancePage.tsx:236]: I changed the policy for loading tabs content (external vs. lazy child)
So, now the structure will look a bit differently. However fixed it: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145/commits/ba9d97a9832a0fb0ad5bb3b4fa58fbb9381b1327

File: src/app/Instance/InstancePage/InstancePage.css [hunk 3]
```diff
@@ -5,3 +5,8 @@
 .instance-page__actions .pf-c-dropdown__menu {
   min-width: 13rem;
 }
+
+#instance-page__tabs-error-handling
+  .instance-page__tabs-error-handling__section {
+  height: 37rem;
+}
```
[the remark this answers is not in the document]
Comment (vpellegrino) [reply src/app/Instance/InstancePage/InstancePage.css]: Good catch. The purpose of this rule was to give a minimum height to the section, when the EH strategy is `Ignore`: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145/commits/a1d010b413eae99fbdcd1ead32823e112ea3e7bd

File: src/app/Instance/InstancePage/ErrorHandlingTabContent.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            </DescriptionListDescription>
+          </React.Fragment>
+        ))}
+      </DescriptionListGroup>
+    );
+  }, [errorHandlingParameters]);
+
+  const getErrorHandlingMethodLabel = useMemo(() => {
+    if (errorHandlingType) {
+      return getErrorHandlingMethodByType(errorHandlingType).label;
```
[the remark this answers is not in the document]
Comment (vpellegrino) [reply src/app/Instance/InstancePage/ErrorHandlingTabContent.tsx]: Renamed: https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox-ui/pull/145/commits/43d1974faff75aeef8a9b2cb2eedb4c74c6ecad0

**Review by kelvah — APPROVED** [review]
Thank you @vpellegrino

**Pull request merged**